benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/HubConnectionReceiveBenchmark.cs

```diff
@@ -0,0 +1,160 @@
+
+                _pipe = new TestDuplexPipe();
+                _pipe.AddReadResult(new ValueTask<ReadResult>(handshakeResponseResult));
+            }
+            finally
+            {
+                MemoryBufferWriter.Return(writer);
+            }
+
+            _nextReadTcs = new TaskCompletionSource<ReadResult>();
+            _pipe.AddReadResult(new ValueTask<ReadResult>(_nextReadTcs.Task));
+
+            IHubProtocol hubProtocol;
+
+            var hubConnectionBuilder = new HubConnectionBuilder();
+            if (Protocol == "json")
+            {
+                hubProtocol = new JsonHubProtocol();
+            }
+            else
+            {
+                hubProtocol = new MessagePackHubProtocol();
+            }
+
+            hubConnectionBuilder.Services.TryAddEnumerable(ServiceDescriptor.Singleton(typeof(IHubProtocol), hubProtocol));
+
+            _invocationMessageBytes = hubProtocol.GetMessageBytes(new InvocationMessage(MethodName, arguments));
+
+            var delegateConnectionFactory = new DelegateConnectionFactory(
+                format =>
+                {
+                    var connection = new DefaultConnectionContext();
+                    // prevents keep alive time being activated
+                    connection.Features.Set<IConnectionInherentKeepAliveFeature>(new TestConnectionInherentKeepAliveFeature());
+                    connection.Transport = _pipe;
+                    return Task.FromResult<ConnectionContext>(connection);
+                },
+                connection =>
+                {
+                    connection.Transport.Output.Complete();
+                    connection.Transport.Input.Complete();
+                    return Task.CompletedTask;
+                });
+            hubConnectionBuilder.Services.AddSingleton<IConnectionFactory>(delegateConnectionFactory);
+
+            _hubConnection = hubConnectionBuilder.Build();
+            _hubConnection.On(MethodName, arguments.Select(v => v.GetType()).ToArray(), OnInvoke);
+            _hubConnection.StartAsync().GetAwaiter().GetResult();
+        }
+
+        private Task OnInvoke(object[] args)
+        {
+            _currentInterationMessageCount++;
+
+            if (_currentInterationMessageCount == MessageCount)
+            {
+                _currentInterationMessageCount = 0;
+                _waitTcs.SetResult(true);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        [Params(0, 1, 10, 100)]
+        public int ArgumentCount;
+
+        [Params(1, 100)]
+        public int MessageCount;
+
+        [Params("json", "messagepack")]
+        public string Protocol;
+
+        [GlobalCleanup]
+        public void GlobalCleanup()
+        {
+            _nextReadTcs.SetResult(new ReadResult(default, false, true));
+            _hubConnection.StopAsync().GetAwaiter().GetResult();
+        }
+
+        public void OperationSetup()
```

Comment:
Make this `[IterationSetup]` so it doesn't count as part of the actual benchmark run.

Reply:
This originally was called for `IterationSetup` but each method call is an operation, and many operations make up an iteration, so it didn't work.

https://github.com/dotnet/BenchmarkDotNet/blob/master/docs/guide/Advanced/SetupAndCleanup.md#iterationsetup

Reply:
Oh interesting :(